File: src/System.Windows.Forms.Analyzers/vb/tests/Analyzer/AvoidPassingTaskWithoutCancellationToken/TestData/ImplicitInvokeAsyncInControlSubclass/AnalyzerTestCode.vb
Option Strict On
Option Explicit On

Imports System
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace VisualBasicControls

    Public Class MyCustomControl
        Inherits Control

        Public Sub TestImplicitInvokeAsyncCalls()
            ' A sync Action delegate is fine
            Dim okAction As New Action(Sub() Text = "Hello, World!")

            ' A sync Func delegate is also fine
            Dim okFunc As New Func(Of Integer)(Function() 42)

            ' Task-returning delegate without CancellationToken - should trigger warning
            Dim notOkAsyncFunc As New Func(Of Task)(Function()
                                                        Text = "Hello, World!"
                                                        Return Task.CompletedTask
                                                    End Function)

            ' This should be flagged as it's using an implicit InvokeAsync call with a Task-returning delegate
            Dim task1 = InvokeAsync(notOkAsyncFunc, CancellationToken.None)

            ' Task<T>-returning delegate without CancellationToken - should trigger warning  
            Dim notOkAsyncFunc2 As New Func(Of Task(Of Integer))(Function()
                                                                     Text = "Hello, World!"
                                                                     Return Task.FromResult(42)
                                                                 End Function)

            ' This should be flagged as it's using an implicit InvokeAsync call with a Task<T>-returning delegate
            Dim task2 = InvokeAsync(notOkAsyncFunc2, CancellationToken.None)

            ' ValueTask-returning delegate without CancellationToken - should trigger warning
            Dim notOkAsyncFunc3 As New Func(Of ValueTask)(Function()
                                                              Text = "Hello, World!"
                                                              Return ValueTask.CompletedTask
                                                          End Function)

            ' This should be flagged as it's using an implicit InvokeAsync call with a ValueTask-returning delegate
            Dim task3 = InvokeAsync(notOkAsyncFunc3, CancellationToken.None)

            ' A proper implementation with CancellationToken - should not trigger warning
            Dim okAsyncFunc As New Func(Of CancellationToken, ValueTask)(Function(ct)
                                                                             Text = "Hello, World!"
                                                                             Return ValueTask.CompletedTask
                                                                         End Function)

            ' This should be fine as it properly handles CancellationToken
            Dim task4 = InvokeAsync(okAsyncFunc, CancellationToken.None)

            ' Another proper implementation with CancellationToken - should not trigger warning
            Dim okAsyncFunc2 As New Func(Of CancellationToken, ValueTask(Of Integer))(Function(ct)
                                                                                          Text = "Hello, World!"
                                                                                          Return ValueTask.FromResult(42)
                                                                                      End Function)

            ' This should be fine as it properly handles CancellationToken
            Dim task5 = InvokeAsync(okAsyncFunc2, CancellationToken.None)

            ' Finally, let's make sure we don't create false positives for non-InvokeAsync methods
            Dim result = SomeOtherMethod(Function() Task.CompletedTask, CancellationToken.None)
        End Sub

        ' This is just to verify we don't create false positives
        Private Function SomeOtherMethod(func As Func(Of Task), ct As CancellationToken) As Task
            Return Task.CompletedTask
        End Function
    End Class

    Public Module Program
        Public Sub Main()
            Dim control As New MyCustomControl()
            control.TestImplicitInvokeAsyncCalls()
        End Sub
    End Module

End Namespace

File: src/System.Windows.Forms.Analyzers/cs/tests/Analyzer/AvoidPassingTaskWithoutCancellationToken/TestData/ImplicitInvokeAsyncInControlSubclass/AnalyzerTestCode.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace CSharpControls;

public class MyCustomControl : Control
{
    public void TestImplicitInvokeAsyncCalls()
    {
        // A sync Action delegate is fine
        var okAction = new Action(() => Text = "Hello, World!");
        
        // A sync Func delegate is also fine
        var okFunc = new Func<int>(() => 42);

        // Task-returning delegate without CancellationToken - should trigger warning
        var notOkAsyncFunc = new Func<Task>(() =>
        {
            Text = "Hello, World!";
            return Task.CompletedTask;
        });

        // This should be flagged as it's using an implicit InvokeAsync call with a Task-returning delegate
        var task1 = InvokeAsync(notOkAsyncFunc, CancellationToken.None);

        // TaskT<T>-returning delegate without CancellationToken - should trigger warning  
        var notOkAsyncFunc2 = new Func<Task<int>>(() =>
        {
            Text = "Hello, World!";
            return Task.FromResult(42);
        });

        // This should be flagged as it's using an implicit InvokeAsync call with a Task<T>-returning delegate
        var task2 = InvokeAsync(notOkAsyncFunc2, CancellationToken.None);

        // ValueTask-returning delegate without CancellationToken - should trigger warning
        var notOkAsyncFunc3 = new Func<ValueTask>(() =>
        {
            Text = "Hello, World!";
            return ValueTask.CompletedTask;
        });

        // This should be flagged as it's using an implicit InvokeAsync call with a ValueTask-returning delegate
        var task3 = InvokeAsync(notOkAsyncFunc3, CancellationToken.None);

        // Should have generated: This is OK, since we're passing a cancellation token.
        var okAsyncFunc = new Func<CancellationToken, ValueTask>((cancellation) =>
        {
            Text = "Hello, World!";
            return ValueTask.CompletedTask;
        });

        //// Had generated: A proper implementation with CancellationToken - should not trigger warning
        //var okAsyncFunc = new Func<CancellationToken, Task>((ct) =>
        //{
        //    Text = "Hello, World!";
        //    return Task.CompletedTask;
        //});

        // This should be fine as it properly handles CancellationToken
        var task4 = InvokeAsync(okAsyncFunc, CancellationToken.None);

        // Another proper implementation with CancellationToken - should not trigger warning
        var okAsyncFunc2 = new Func<CancellationToken, ValueTask<int>>((ct) =>
        {
            Text = "Hello, World!";
            return ValueTask.FromResult(42);
        });

        // This should be fine as it properly handles CancellationToken
        var task5 = InvokeAsync(okAsyncFunc2, CancellationToken.None);

        // Finally, let's make sure we don't create false positives for non-InvokeAsync methods
        var result = SomeOtherMethod(() => Task.CompletedTask, CancellationToken.None);
    }

    // This is just to verify we don't create false positives
    private Task SomeOtherMethod(Func<Task> func, CancellationToken ct)
    {
        return Task.CompletedTask;
    }
}

public static class Program
{
    public static void Main()
    {
        var control = new MyCustomControl();
        control.TestImplicitInvokeAsyncCalls();
    }
}
